### ISSUE TITLE
pylance reports not accessed for "private" functions used elsewhere 

### ISSUE THREAD

**Issue by VolkerH** [microsoft/pylance-release#5755]
## Environment data

Pylance async client (2024.4.1) started with python extension (2024.4.1)
Pylance language server 2024.4.1 (pyright version 1.1.356, commit 6652c4a8)
Windows 10 64bit
Python 3.11.0

## Code Snippet

Here are two screenshots that show everything that's needed to understand the issue:

![image](https://github.com/microsoft/pylance-release/assets/2210044/9c228676-8260-4c20-95da-b06a2f260475)

![image](https://github.com/microsoft/pylance-release/assets/2210044/f72456ef-eea0-401c-b44e-7d29fb762de0)

## Repro Steps

1. define a function marked as private (with a name starting  with `_` by convention) in one `.py` file but don't use it in the same file
1. import the function defined in the previous step from another `.py` file and make use of it there !

## Expected

The function `_not_used_here_but_elsewhere` should be rendered in saturated colour and not be reported as not being accessed as it is accessed in `user_of_private_function.py`.
While one should arguably not import private functions elsewhere this is not a language requirement. In some python projects, private as in starting with `_` might mean "not part of the public API" as opposed to "only use in this namespace".

## Actual behavior

The function `_not_used_here_but_elsewhere` is erroneously reported as not being accessed and rendered in de-saturated colour.

## EDIT: to add additional comments

* even if one might argue that one shouln't import private functions from another module, the error message is factually incorrect, as the function is actually accessed. The current message may mislead a user to remove the apparently unused code, thus breaking something inadvertendly.
* As mentioned a leading `_` meaning private is a convention, not a language rule. I am not sure the convention establishes whether the notion of private (in the sense of "not part of the public API") can only apply to the module level and not the package level.

EDIT2:

* ideally, this would be flagged as `Defined as private but accessed outside module` with a rule that can be disabled.
* meanwhile realised this is a duplicate of https://github.com/microsoft/pyright/issues/6747 , but I would urge to re-open this. Disabling all reporting of code that is not accessed generally (i.e. not just for this case of the private function) is not a desirable solution.
* the behaviour is inconsistent with VS Codes find all references:
![image](https://github.com/microsoft/pylance-release/assets/2210044/90a75622-3a5a-4983-a624-8706c04946c6)
* PyCharm, which also has a very good language server, does not show this as unused code:
![image](https://github.com/microsoft/pylance-release/assets/2210044/918b4890-e7f2-4549-b658-662a5122deaa)
with a small change to the import statement (that doesn't make a difference in PyLance/VSCode) it reports the breach of convention where it is commited 
![image](https://github.com/microsoft/pylance-release/assets/2210044/370adef4-75ee-47fc-99d0-8f2aa48a99ac)

**Comment by erictraut** [comment]
You are importing a private function into another module. Private functions are assumed to be private to the module in which they are defined (hence the name "private"). If your intent is to allow a symbol to be imported by another module, you should name it without a leading underscore. Pylance will then assume that it might be referenced by another module and not mark it as unreferenced.

**Comment by VolkerH** [comment]
Thanks. I think I made clear I understand that one should ideally not import private functions.
However I think the following points (mostly reiterating above) are nevertheless worth considering:

* the message flagged here is factually incorrect. 
* the message is misleading
* it is inconsistent with "find all references" (not a pyright thing as far as I understand) but in terms of VS Code it is inconsistent
* the `_` meaning private is a convention, not a language specification
* the message is not (by default, I understand this can be configured) raised in the place where the convention is breached

**Comment by erictraut** [comment]
Pyright is working as intended here. This isn't a bug.

When computing whether a symbol is accessed elsewhere, a static analyzer needs to make assumptions about the "global visibility" of the symbol. By default, all module-level symbols in Python are globally visible to any other module (regardless of whether those modules are part of your project). The Python language has no way to prevent a module-level symbol from being accessed from another module. (Technically, there's no way to prevent a class-scoped or function-scoped variable from being accessed either, since those are available at runtime through introspection.) So, a static type checker needs to rely on language conventions. In this case, it assumes that if the author of a module names a module-level symbol as "private", the intent is that it is not globally visible to other modules. Therefore, if it's not accessed anywhere within its own module, it is considered unreferenced.

If you want pyright to emit a diagnostic for inappropriately importing a private member from another module, you can enable the `reportPrivateUse` diagnostic check.

**Comment by VolkerH** [comment]
>  you should name it without a leading underscore. 

For me, the value of a tool like Pylance is to help me uncover parts of the code base where I didn't do something the way should have done it. That is not the case here for me.

In this case, I would be less confused if the issue was raised where the private symbol is imported  (`user_of_private_function.py`) or instead if the code in the function definition in `private_functions.py` was reported as _"defined as private but accessed from another module"_  at the place where it is currently reported as not accessed.

**Comment by VolkerH** [comment]
>  In this case, it assumes that if the author of a module names a module-level symbol as "private", the intent is that it is not globally visible to other modules. Therefore, if it's not accessed anywhere within its own module, it is considered unreferenced.

I can understand your reasoning that the error message is correct as in "not accessed in this module" .
However, in the context of IDEs such as VS Code, I would think most people work on projects or larger codebases.  Personally, I would like to know where it is imported anywhere in the code base, and I think I am not alone with that opinion (see e.g. Pycharm and the upvotes on the prior issue in pyright which I discovered only after opening this one).

Anyway. I can see your point and I have no more to add to my points.

**Comment by erictraut** [comment]
Pyright performs local analysis of Python source files. It doesn't perform global analysis across all source files. That would be computationally prohibitive. It would also be impossible in the general case because it would require pyright to know about all source files — that that are present and not present within your project — that could potentially import from any module within your project.

Without global analysis, it's not possible to report "defined as private but accessed from another module". The error must be detected at the point where the violation occurs, hence the `reportPrivateUsage` check. This check isn't enabled by default, but you can opt into it if you would like pyright to flag such errors.

Pyright doesn't generate diagnostics for unreferenced module-level variables even if they are marked as private. Rather than generating a diagnostic (an error or warning), it uses a "hint" — a subtle gray effect — to indicate that a private module-level variable is not referenced anywhere within the file. This isn't a bug per-se, which is why it doesn't generate a diagnostic. However, it is potentially suspicious. You can choose to ignore this condition if it's something you intended.
